hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/CastMap.java

```diff
@@ -165,21 +192,132 @@ void add(int pos, LogicalType fromType, LogicalType toType) {
+   *
+   * @param row      Non-null row data to be converted; however, fields might contain nulls
+   * @param fromType The input LogicalType of the row data to be converted from
+   * @param toType   The output LogicalType of the row data to be converted to
+   * @return Converted row that has the structure/specifications of that defined by the output LogicalType
+   */
+  private static RowData doRowConversion(@Nonnull RowData row, LogicalType fromType, LogicalType toType) {
+    // note: InternalSchema.merge guarantees that the schema to be read fromType is orientated in the same order as toType
+    // hence, we can match types by position as it is guaranteed that it is referencing the same field
+    List<LogicalType> fromChildren = fromType.getChildren();
+    List<LogicalType> toChildren = toType.getChildren();
+    ValidationUtils.checkArgument(fromChildren.size() == toChildren.size(),
+        "fromType [" + fromType + "] size: != toType [" + toType + "] size");
+
+    GenericRowData rowData = new GenericRowData(toType.getChildren().size());
+    for (int i = 0; i < toChildren.size(); i++) {
+      Object fromVal = RowData.createFieldGetter(fromChildren.get(i), i).getFieldOrNull(row);
+      Object toVal;
+
+      // need to handle nulls to prevent NullPointerException in #getConversion()
+      if (fromChildren.get(i).equals(toChildren.get(i)) || fromVal == null) {
+        // no conversion required if the types are the same / fromVal is null
+        toVal = fromVal;
+      } else {
+        // conversion required
+        toVal = getConversion(fromChildren.get(i), toChildren.get(i)).apply(fromVal);
+      }
+
+      rowData.setField(i, toVal);
+    }
+    return rowData;
   }
 
-  private DecimalData toDecimalData(Number val, LogicalType decimalType) {
+  /**
+   * Helper function to convert a Number object to Flink's DecimalData format.
+   *
+   * @param val         Number object to be converted
+   * @param decimalType DecimalType containing the output decimal's precision and scale specifications
+   * @return Converted decimal that is wrapped in Flink's DecimalData object.
+   */
+  private static DecimalData toDecimalData(@Nonnull Number val, LogicalType decimalType) {
     BigDecimal valAsDecimal = BigDecimal.valueOf(val.doubleValue());
     return toDecimalData(valAsDecimal, decimalType);
   }
 
-  private DecimalData toDecimalData(BigDecimal valAsDecimal, LogicalType decimalType) {
+  /**
+   * Helper function to convert a BigDecimal object to Flink's DecimalData format.
+   *
+   * @param valAsDecimal BigDecimal object to be converted
+   * @param decimalType  DecimalType containing the output decimal's precision and scale specifications
+   * @return Converted decimal that is wrapped in Flink's DecimalData object.
+   */
+  private static DecimalData toDecimalData(@Nonnull BigDecimal valAsDecimal, LogicalType decimalType) {
     return DecimalData.fromBigDecimal(
```

Comment:
By default all the params should be non null, there is no need to add annotation.